PR: remove dependency on walkdir and support loading from specified file extensions, falling back on .js

fluffynuts opened the pull request:
Hi

We have been doing some work in Typescript with a client. They introduced cnfg as the mechanism for building and applying configuration based on small modules and using the CNFG_\* environment overrides. Unfortunately, we found that we often had loading issues under Windows -- sometimes configuration would just "not be there" according to cnfg, though the files were on disk.

The problem was tracked down to the dependent package walkdir, which uses filesystem inode identifiers as a key when keeping a lookup of files whilst traversing the filesystem. It turns out that some filesystems (of which NTFS is a prime example), may not guarantee unique inode identifiers when getting results from node's fs.lstat() and fs.lstatSync() functions. I've replicated the issue myself, but this is a long-standing problem that I've found scattered across the web, with a really old example here: [https://github.com/mhevery/jasmine-node/issues/378](https://github.com/mhevery/jasmine-node/issues/378)

Probably a good thing to do would be to PR against walkdir too with a fix for the issue. However, another useful change would be to allow loading from files other than just .js files -- for example when running under an on-the-fly transpiler like ts-node. This is also included, at minimal cost.

Please review the commit message for more information. 

I had worked around the load issue with a very minimal file loader which did much the same as cnfg, but it turns out our client also relies on the CNFG_\* variables. Rather than re-implement that functionality too, I decided to fork and provide a PR, if you're interested. However, I need to release to npmjs.com so I can use in production code today. If you accept the PR, I'll gladly switch back to your package and request npmjs.com to take down mine, which I'm (unimaginatively) calling cnfg-ng (next-gen, since I can't put ++ in the package name)

Commit:
Commit for PR:
- remove dependency on walkdir, which doesn't always behave as expected
    due to its design around the assumption that filesystem inodes
    always have unique identifiers when queried with node's fs module.
    - this is not always true on certain filesystems, NTFS being a prime
        example, where the inode number appears to be related to file
        creation time, such that the output from a transpilation process
        (such as typescript) can (and does) result in files with the
        same inode number, resulting in some files not being reported
        as existing when using walkdir (and hence some config files
        being ignored by cnfg)
        - the walkdir motivation appears to be to ignore multiple
            hardlinks of the same file, which would be bad practice
            for configuration and would, at any rate, simply result
            in duplicated sections of configuration with different names
- also added the ability to specify a selection of file extensions to
  examine for loading, which makes cnfg useful under
  transpile-on-the-fly environments like ts-node
Minor changes:
- all tests are replicated to prove the code working under ts-node
- updated the test mechanism to not depend on Make

Commit:
* fix derp in package.json
    - prepublish should "run-s test", not just "test"

fluffynuts commented:
Sorry, this fixes a derp in the package.json

boo1ean commented:
Hi!
Big thanks for investigation, you did a great job.
Unfortunately I don't have enough time to merge this PR today (need to test it, cause we use it in all prod projects).

Note: I would like to keep original code style:
- identations = tabs
- variables definitions = `var` per line

fluffynuts commented:
Hi

Sorry, I didn't realise about the tabs vs spaces -- spaces is just how we
normally do things here, but of course, I have no problem with using tabs
for the PR. The "one var statement" is just a habit of mine -- again, not,
by any means, essential.

Should I fix these and update the PR? I'll still have to release (have
already, in fact) a fork package (cnfg-ng) because the client is getting
quite excited about fixing this NOW as they'd like to deploy to production
at start of their day tomorrow (they're in AU, we're in ZA, so I think that
means they want to deploy in about 12 hours). I can quite easily revert
back to your package when you're happy with the PR and have done a release.

-d

On 8 August 2016 at 11:37, Egor Gumenyuk <notifications@github.com> wrote:

> Hi!
> Big thanks for investigation, you did a great job.
> Unfortunately I don't have enough time to merge this PR today (need to
> test it, cause we use it in all prod projects).
>
> Note: I would like to keep original code style:
>
>    - identations = tabs
>    - variables definitions = var per line
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boo1ean/cnfg/pull/3#issuecomment-238188000>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEc_nOpQDZLGZVCoDB1MVhekiAlFwBb9ks5qdvjZgaJpZM4Je1Oj>
> .
>



-- 
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
If you say that getting the money is the most important thing
You will spend your life completely wasting your time
You will be doing things you don't like doing
In order to go on living
That is, to go on doing things you don't like doing

Which is stupid.

- Alan Watts
https://www.youtube.com/watch?v=-gXTZM_uPMY

*Quidquid latine dictum sit, altum sonatur. *

Commit:
Add .editorconfig to set tabs instead of spaces, as preferred by author
- update files using double-spaces for indentation to use tabs
- update files having multiple declarations per "var" statement
	to have one declaration per "var"

boo1ean commented:
Thanks!
But I don't have much time today for reviewing whole changes. I suggest you continue with your solution for now or wait for tomorrow.

fluffynuts commented:
Hi

I have -- I've submitted the changes for your review and released an
interim package (cnfg-ng) which the project is using.

-d

On 8 August 2016 at 15:17, Egor Gumenyuk <notifications@github.com> wrote:

> Thanks!
> But I don't have much time today for reviewing whole changes. I suggest
> you continue with your solution for now or wait for tomorrow.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boo1ean/cnfg/pull/3#issuecomment-238232027>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEc_nGWjgMWibqMDSS3vwn0IVQMk4IAiks5qdyyGgaJpZM4Je1Oj>
> .
>



-- 
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
If you say that getting the money is the most important thing
You will spend your life completely wasting your time
You will be doing things you don't like doing
In order to go on living
That is, to go on doing things you don't like doing

Which is stupid.

- Alan Watts
https://www.youtube.com/watch?v=-gXTZM_uPMY

*Quidquid latine dictum sit, altum sonatur. *

fluffynuts commented:
Just a ping to ask if you can find the time to review this PR -- I'd really like to drop my forked package (:

fluffynuts commented:
ping-back: I'd like to use this in a new project and would really like to make a choice about config loading at the outset. I can just use cnfg-ng (my package), but I'd really prefer to use the proper upstream. Any thoughts?

boo1ean commented:
Hi, sorry for ignoring this for a while.

I've couple of questions:

in `package.json` we have postinstall script `typings install` we do we need it?

fluffynuts commented:
Hi

I think that might be an accidental insert in my part - the project I was
consuming cnfg from was a TypeScript project. I'm quite sure there's no
TypeScript in my PR, so I'd happily say to remove that (or I can, and the
PR can be updated, if you prefer).

-d

On Sep 14, 2016 10:47, "Egor Gumenyuk" <notifications@github.com> wrote:

> Hi, sorry for ignoring this for a while.
>
> I've couple of questions:
>
> in package.json we have postinstall script typings install we do we need
> it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boo1ean/cnfg/pull/3#issuecomment-246946662>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEc_nImUO4CHLh66zb90IhLIA-2XkAv2ks5qp7SpgaJpZM4Je1Oj>
> .
>

boo1ean commented:
Got it.
I'll take care, thanks!

boo1ean commented:
I've crated pull request based on this with some small fixes PR #4

boo1ean commented:
Published `cnfg@1.2.0`, you can check it out :)

fluffynuts commented:
Thanking you sir! (:

I shall deprecate cnfg-ng asap (:

-d

On 14 September 2016 at 14:33, Egor Gumenyuk notifications@github.com
wrote:

> Published cnfg@1.2.0, you can check it out :)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/boo1ean/cnfg/pull/3#issuecomment-246998410, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AEc_nHegy0zqpY0lzmvNOmgnIKRJtr0vks5qp-mngaJpZM4Je1Oj
> .

## 

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
If you say that getting the money is the most important thing
You will spend your life completely wasting your time
You will be doing things you don't like doing
In order to go on living
That is, to go on doing things you don't like doing

Which is stupid.
- Alan Watts
  https://www.youtube.com/watch?v=-gXTZM_uPMY

*Quidquid latine dictum sit, altum sonatur. *
